sw/device/lib/runtime/pmp.h

```diff
@@ -0,0 +1,312 @@
+   */
+  kPmpRegionConfigureBadAddress,
+  /**
+   * The requested region was not configured correctly. From the "Volume II:
+   * RISC-V Privileged Architectures V20190608-Priv-MSU-Ratified":
+   * "Implementations will not raise an exception on writes of unsupported
+   * values to a WARL field. Implementations can return any legal value on the
+   * read of a WARL field when the last write was of an illegal value, but the
+   * legal value returned should deterministically depend on the illegal
+   * written value and the value of the field prior to the write"
+   */
+  kPmpRegionConfigureWarlError,
+} pmp_region_configure_result_t;
+
+/**
+ * Disables PMP address matching.
+ *
+ * Address matching is disabled for `region`, however the corresponding
+ * pmpaddr can still be used as a TOR range start address. For example, when
+ * pmpcfgN is set to TOR, but pmpcfgN-1 is set to OFF, then pmpaddrN-1 can
+ * be used as TOR range start address.
+ *
+ * Please see the PMP specification reference and general information at the
+ * top of this header file.
+ *
+ * Note: this function will clears PMP `region` configuration.
+ *
+ * @param region PMP region to configure and set address for.
+ * @param address PMP address to be set.
+ * @return `pmp_region_configure_result_t`.
+ */
+PMP_WARN_UNUSED_RESULT
+pmp_region_configure_result_t pmp_region_configure_off(
+    pmp_region_index_t region, pmp_region_address_t address);
+
+/**
+ * PMP region access and address NA4 configuration result.
+ */
+typedef enum pmp_region_configure_na4_result {
+  kPmpRegionConfigureNa4Ok = kPmpRegionConfigureOk,
+  kPmpRegionConfigureNa4Error = kPmpRegionConfigureError,
+  kPmpRegionConfigureNa4BadArg = kPmpRegionConfigureBadArg,
+  kPmpRegionConfigureNa4BadRegion = kPmpRegionConfigureBadRegion,
+  kPmpRegionConfigureNa4BadAddress = kPmpRegionConfigureBadAddress,
+  kPmpRegionConfigureNa4WarlError = kPmpRegionConfigureWarlError,
+  /**
+   * NA4 is not available with granularity "G" > 0. Please see:
+   * "Volume II: RISC-V Privileged Architectures V20190608-Priv-MSU-Ratified",
+   * "3.6 Physical Memory Protection", "Table 3.10" and section
+   * "Address Matching".
+   */
+  kPmpRegionConfigureNa4Unavailable,
+} pmp_region_configure_na4_result_t;
+
+/**
+ * Configures PMP address matching to Naturally aligned four-byte (NA4).
+ *
+ * When pmpcfgN is set to NA4 then the protected region range is formed from
+ * pmpaddrN and extends to the next 4 bytes.
+ *
+ * Please see the PMP specification reference and general information at the
+ * top of this header file.
+ *
+ * @param region PMP region to configure and set the address for.
+ * @param config PMP region configuration information.
+ * @param address PMP address to be set.
+ * @return `pmp_region_configure_na4_result_t`.
+ */
+PMP_WARN_UNUSED_RESULT
+pmp_region_configure_na4_result_t pmp_region_configure_na4(
+    pmp_region_index_t region, const pmp_region_config_t *config,
+    pmp_region_address_t address);
```

Comment:
I suggested this was combined with `pmp_region_configure_napot`, as you can choose the logic of how to encode the correct setting based on the `size` parameter. This means a simpler interface, where callers don't necessarily have to care what their size is, as long as it's larger than 4-bytes (rather than a specific call for 4-bytes, and other calls for 8-bytes). What was your opinion on this suggestion?

Reply:
I will give it a try next time I am working on the PMP.